src/stream/src/executor/backfill/arrangement_backfill.rs

```diff
@@ -358,20 +361,60 @@ where
+                                        // Because skipping over long tombstone ranges is a transient step,
+                                        // once we skip over the initial run, we don't need to do it again.
+                                        //
+                                        // If we hit this code section, it means snapshot read is really slow.
+                                        // Even if we yielded a record here, the backfill progress will still be slow,
+                                        // since we only return one record per epoch.
+                                        // Consider barrier interval at 1s.
+                                        // And upstream has 1M records.
+                                        // So we take 1M seconds to finish backfill, which is 11.57 days.
+                                        // So that's basically the same as never finishing backfill.
+                                        if !yielded {
+                                            let builder = builders.get_mut(&vnode).unwrap();
+                                            if let Some(chunk) = builder.append_one_row(row) {
+                                                yield Message::Chunk(Self::handle_snapshot_chunk(
+                                                    chunk,
+                                                    vnode,
+                                                    &pk_in_output_indices,
+                                                    &mut backfill_state,
+                                                    &mut cur_barrier_snapshot_processed_rows,
+                                                    &mut total_snapshot_processed_rows,
+                                                    &self.output_indices,
+                                                )?);
+                                            }
+                                            yielded = true;
+                                        } else {
+                                            let new_pos = row.project(&pk_in_output_indices);
+                                            assert_eq!(new_pos.len(), pk_in_output_indices.len());
+                                            backfill_state.update_progress(
+                                                vnode,
+                                                new_pos.to_owned_row(),
+                                                false,
+                                                0,
+                                            )?;
+                                        }
 
-                                    break;
+                                        remaining_vnodes.set_bit(vnode_idx);
+                                        if remaining_vnodes.is_empty() {
+                                            break;
+                                        }
```

Comment:
After it set a bit, could it be empty?

Reply:
It should be set to 0 instead. Let me fix it later.

Reply:
https://github.com/risingwavelabs/risingwave/pull/17266/commits/b29adfd7b1c1836b8198d36368209743d27e9dab Fixed.

Since `Bitmap` is designed to be an immutable data structure (as per its docs), it is hard to update correctly.
I have opted to convert it to a `BitSet` instead.